packages/uilib/src/lib/plugins/type-designer/editor-events/editor-event-handler.ts

```diff
@@ -0,0 +1,87 @@
+import { CreateBayEvent, CreateIEDEvent, CreateLDeviceEvent, CreateSubstationEvent, CreateVoltageLevelEvent } from "./event-types"
+import type { Create, Delete } from "./editor-events"
+import type { BayType, IEDType, LDeviceType, VoltageLevelType, SubstationType } from "../types"
+import { getNodeName } from "../components";
+
+export class EditorEventHandler {
+    private readonly editorActionName = "editor-action"
+
+    constructor(private readonly root: HTMLElement, private readonly dataTemplates: Element) {
+    }
+
+    public dispatchCreateBay(event: CreateBayEvent): void {
+        this.dispatchCreate(event);
+    }
+
+    public dispatchCreateSubstation(event: CreateSubstationEvent): void {
+        this.dispatchCreate(event);
+    }
+
+    public dispatchCreateLDevice(event: CreateLDeviceEvent): void {
+        this.dispatchCreate(event);
+    }
+
+    public dispatchCreateIED(event: CreateLDeviceEvent): void {
+        this.dispatchCreate(event);
+    }
+
+    public dispatchCreateVoltageLevel(event: CreateLDeviceEvent): void {
+        this.dispatchCreate(event);
+    }
+
+    private dispatchCreate(event: CreateBayEvent | CreateIEDEvent | CreateVoltageLevelEvent | CreateLDeviceEvent | CreateSubstationEvent): void {
+        const newNode = this.buildNewNode(event.type);
+        const replaces = this.buildCreateEvents(newNode);
+        const combinedEditorEvent = this.buildEditorActionEvent(replaces);
+        this.root.dispatchEvent(combinedEditorEvent);
+    }
+
+    private buildCreateEvents(newElement: Element): Create[] {
+        return [{
+            new: { parent: this.dataTemplates, element: newElement },
+        }];
+    }
+
+    private buildNewNode(node: BayType | IEDType | LDeviceType | VoltageLevelType | SubstationType): Element {
+        const nodeName = getNodeName(node);
+        const xmlString = this.createXmlString(nodeName, node);
+        const parser = new DOMParser();
+        const doc = parser.parseFromString(xmlString, "application/xml");
+        return doc.documentElement;
+    }
+
+    private createXmlString(nodeName: string, node: BayType | IEDType | LDeviceType | VoltageLevelType | SubstationType): string {
+        const attributes = Object.entries(node)
+            .map(([key, value]) => `${key}="${value}"`)
+            .join(" ");
+        return `<${nodeName} ${attributes}></${nodeName}>`;
+    }
```

Comment:
Is there any reason for manually building the Xml string instead of using `document.createElement` and setting its attributes?

Reply:
For some reason it doesn't work for me. I tried to find the reason why it would assign a wrong name, but the best i got was this format:

<ldevicetype xmlns="http://www.w3.org/1999/xhtml" inst="sdfa" desc="awerxc" id="234"> </ldevicetype>

Keeping it open for now

Reply:
Open SCD does have a package for doing XML stuff https://github.com/openscd/open-scd/blob/main/packages/xml/src/foundation.ts, ideally we would use that, but I am not sure if it is imported at the moment.

For now we could copy the `createElement` method and put it in `packages/core/xml`

Reply:
Use of OpenSCD API